### ISSUE TITLE
How do we use flash-attn?

### ISSUE THREAD

**Issue by DarkAlchy** [kijai/ComfyUI-LuminaWrapper#5]
Compiled it and there is no way, or is there, that this should be this slow.  2it/s on a 4090?

**Comment by kijai** [comment]
I'm getting 2.92 it/s at 1024x1024 with 4090 when using flash attention, so yeah it's bit slow. Without it you'd be looking at seconds per iteration instead, so it does seem to be working if you are using a higher resolution.

**Comment by DarkAlchy** [comment]
Alright, now the quality I just showed discord and I am stumped.

![image](https://github.com/kijai/ComfyUI-LuminaWrapper/assets/10561690/c2c3906f-c2af-4421-b15b-3e01d332bc58)

**Comment by patientx** [comment]
So this is why I was getting around 30-40 sec / it on rx 6600 :) we don't have a way to flash attention at all here so good luck to nvidia brethren.

**Comment by festivus37** [comment]
> I'm getting 2.92 it/s at 1024x1024 with 4090 when using flash attention, so yeah it's bit slow. Without it you'd be looking at seconds per iteration instead, so it does seem to be working if you are using a higher resolution.

Man, I just can't seem to get it installed. Every time I try "pip install <pasted link to one of the whl files" it just keeps saying ERROR: flash_attn-2.5.9.post1+cu122torch2.3.1cxx11abiFALSE-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.".  I've tried them all.  I'm on windows 11 with (2.3.1+cu121) torch.

**Comment by kijai** [comment]
> > I'm getting 2.92 it/s at 1024x1024 with 4090 when using flash attention, so yeah it's bit slow. Without it you'd be looking at seconds per iteration instead, so it does seem to be working if you are using a higher resolution.
> 
> Man, I just can't seem to get it installed. Every time I try "pip install <pasted link to one of the whl files" it just keeps saying ERROR: flash_attn-2.5.9.post1+cu122torch2.3.1cxx11abiFALSE-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.". I've tried them all. I'm on windows 11 with (2.3.1+cu121) torch.

What's your Python version? That file is "cp39" which would be for Python 3.9.

**Comment by festivus37** [comment]
> > > I'm getting 2.92 it/s at 1024x1024 with 4090 when using flash attention, so yeah it's bit slow. Without it you'd be looking at seconds per iteration instead, so it does seem to be working if you are using a higher resolution.
> > 
> > 
> > Man, I just can't seem to get it installed. Every time I try "pip install <pasted link to one of the whl files" it just keeps saying ERROR: flash_attn-2.5.9.post1+cu122torch2.3.1cxx11abiFALSE-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.". I've tried them all. I'm on windows 11 with (2.3.1+cu121) torch.
> 
> What's your Python version? That file is "cp39" which would be for Python 3.9.

I'm on Python 3.10.11 (windows). so that nugget you mentioned helped a lot. That said I tried every cp10 with the right cuda 1.21 that I had, and every one blew up my various comfy nodes (extra samplers etc), so eventually I gave up. I found a relatively fast workflow at 2.40:1 aspect ratio that renders in about 21 seconds with sdxl refiner. Thanks for the work on these nodes. Enjoy a picture. :)
![ComfyUI_temp_lpnci_00003_](https://github.com/kijai/ComfyUI-LuminaWrapper/assets/170977426/d7efdbcd-97a8-4458-b129-93cde987b885)

**Comment by diodiogod** [comment]
How can I check what is my python environment in my ComfyUi installation to install flahs-attn?

**Comment by sdk401** [comment]
Ok, stupid question, how do I know if I've installed correct flash-attn and it works? I've ran pip install "correct_wheel", it said intalled succesfully, but I'm running comfyui through stability matrix and it has it's own python, so I don't understand if I installed it to my global installation or to the local python in matrix.

Getting pretty long gen times, but no errors, is there some message in the console which shows the mode it's working in? Don't see any errors.

Sampling time looks like this:

ODE Sampling: 100%|██████████| 48/48 [02:08<00:00,  2.67s/it]

2+ minutes for 25 steps on 1024x1024 image, on 8gb 3070. I'm not going into low ram mode, it sits at about 7.7gb of VRAM.
Maybe it's working as intended? Hard to say. Pixart, sd3 and sdxl are sampling about 2-3 times faster for the same steps and resolution.

**Comment by kijai** [comment]
That does sound like it's not working, I'll add console message to make it
clearer. I don't actually use stability matrix so I don't know how it has
the python environment setup, it definitely doesn't use your global python
though.

**Comment by sdk401** [comment]
> definitely doesn't use your global python though.

yeah, i've tried installing directly with python.exe in local python folder, it said "already intalled".
trying one more time with "--force reinstall", I hope it will not break verything else :)

**Comment by sdk401** [comment]
ok, now I see I have cu121, and the file says cu122.
but I can't understand how to instal pytorch with cu122, their configurator does not have it:

![image](https://github.com/kijai/ComfyUI-LuminaWrapper/assets/162196217/14842480-395b-4fc7-b6c3-e6dae7e791f9)

**Comment by sdk401** [comment]
Found the button to install the wheel directly in matrix interface, looks like it worked, time is down to 1+ minute, looks like it's as fast as I'll get on my gpu. Sorry for the comment spam, thanks a lot fot the node :)

ODE Sampling: 100%|██████████| 48/48 [01:13<00:00,  1.54s/it]

**Comment by kijai** [comment]
> How can I check what is my python environment in my ComfyUi installation to install flahs-attn?

This depends what kind of comfy install you have, it generally should tell you in the console log when you start Comfy. 

Manually without running comfy, if it's the portable you can run this to show pytorch version:
`python_embeded\python.exe -m pip show torch`
Python version:
`python_embeded\python.exe --version`

**Comment by ChibiChubu** [comment]
![image_2024-06-18_17-12-19](https://github.com/kijai/ComfyUI-LuminaWrapper/assets/168059770/4e34055f-b66e-41b2-8ccc-fc1c86fccdc4)

what i miss?

**Comment by kijai** [comment]
> ![image_2024-06-18_17-12-19](https://private-user-images.githubusercontent.com/168059770/340616151-4e34055f-b66e-41b2-8ccc-fc1c86fccdc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MDI2MDMsIm5iZiI6MTcxODcwMjMwMywicGF0aCI6Ii8xNjgwNTk3NzAvMzQwNjE2MTUxLTRlMzQwNTVmLWI2NmUtNDFiMi04Y2NjLWZjMWM4NmZjY2RjNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxOFQwOTE4MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYzQyNDY4ODlhZWM5MmZjYzJmYTU5NjBkOGY3ODJhYzYyNWM5MzMyZWM0YzFjMDVlMzA3ODliYTdiNmFmYzc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WHfjO-LXMxmXAtb4FALyKUa6AUc2dvNdQlgyRnhXluI)
> 
> what i miss?

Probably missing one of the dependencies from the requirements.txt

**Comment by DarkAlchy** [comment]
Would I gain anything by going flash-attn2?

**Comment by kijai** [comment]
> Would I gain anything by going flash-attn2?

As far as I understood "flash attention 2" is just flast_attn versioning, as in 2.5.9 is flash attention 2 already.

**Comment by DarkAlchy** [comment]
I only ask because I built it and some other package complained that it was 1.
